src/casts/specify-representation.rkt

```diff
@@ -143,541 +156,339 @@ but a static single assignment is implicitly maintained.
+                 (sr-tagged-array-set! t-tmp-var TYPE-FN-TAG (sr-plus TYPE-FN-FMLS-OFFSET i-var) t-glb-var))
+                t-tmp-var))))
+      (add-new-code! (cons mk-fn-type-glb mk-fn-type-glb-c))
+      (set-box! mk-fn-type-glb-code-label? mk-fn-type-glb-label)
+      mk-fn-type-glb-label)
+    (let ([cl? (unbox mk-fn-type-glb-code-label?)])
+      (or cl? (make-code! mk-tglb))))
 
   (: get-mk-fn-crcn! (Uid -> (Code-Label Uid)))
   (define (get-mk-fn-crcn! mk-crcn)
     (: make-code! (Uid -> (Code-Label Uid)))
     (define (make-code! mk-crcn)
-      (define mk-fn-u (next-uid! "make-fn-coercion"))
-      (define t1-u    (next-uid! "fn-type1"))
-      (define t2-u    (next-uid! "fn-type2"))
-      (define l-u     (next-uid! "label"))
-      (define i-u     (next-uid! "index"))
-      (define a-u     (next-uid! "fn-type-arity"))
-      (define t1r-u   (next-uid! "fn-type1-return"))
-      (define t2r-u   (next-uid! "fn-type2-return"))
-      (define t1a-u   (next-uid! "fn-type1-argument"))
-      (define t2a-u   (next-uid! "fn-type2-argument"))
-      (define c1-u    (next-uid! "resulting_coercion"))
-      (define c2-u    (next-uid! "resulting_coercion"))
-      (define cr-u    (next-uid! "return-coercion"))
-      (define ca-u    (next-uid! "argument-coercion"))
-      (define st-u    (next-uid! "second-tagged"))
-      (define t1      (Var t1-u))
-      (define t2      (Var t2-u))
-      (define l       (Var l-u))
-      (define c1      (Var c1-u))
-      (define c2      (Var c2-u))
-      (define i       (Var i-u))
-      (define a       (Var a-u))
-      (define st      (Var st-u))
-      (define mk-fn   (Code-Label mk-fn-u))
-      (define mk-c    (Code-Label mk-crcn))
-      (define mk-fn-c : D0-Code
-        (Code `(,t1-u ,t2-u ,l-u ,i-u ,a-u)
-          (If (Op '= `(,i ,a))
+      (define-vars mk-fn-crcn t1 t2 l i arity t1r t2r t1a t2a crcn tmp-crcn cr ca tagged-arity)
+      (define mk-fn-crcn-label (Code-Label mk-fn-crcn))
+      (define mk-crcn-label (Code-Label mk-crcn))
+      (define mk-fn-crcn-c : D0-Code
+        (Code `(,t1 ,t2 ,l ,i ,arity)
+          (If (Op '= `(,i-var ,arity-var))
               (Begin
                 (list
-                 (Assign t1r-u (sr-tagged-array-ref t1 TYPE-FN-TAG TYPE-FN-RETURN-INDEX))
-                 (Assign t2r-u (sr-tagged-array-ref t2 TYPE-FN-TAG TYPE-FN-RETURN-INDEX))
-                 (Assign cr-u  (App-Code mk-c `(,(Var t1r-u) ,(Var t2r-u) ,l)))
-                 (Assign c1-u  (Op 'Alloc (list (sr-plus a (Quote 2)))))
-                 (Assign st-u  (Op '+ (list (Op '%<< (list a COERCION-SECOND-TAG-SHIFT))
-                                            COERCION-FN-SECOND-TAG)))
-                 (sr-array-set! c1 COERCION-FN-ARITY-INDEX st)
-                 (sr-array-set! c1 COERCION-FN-RETURN-INDEX (Var cr-u)))
-                (sr-tag-value c1 COERCION-MEDIATING-TAG))
+                 (Assign t1r (sr-tagged-array-ref t1-var TYPE-FN-TAG TYPE-FN-RETURN-INDEX))
```

Comment:
This code could be further improved by use of the begin$ macro which allocates vars with the help of the assign$ macro.

Reply:
I will implement it using `begin$` macro when I move it from `specify-representation`.